dolphinscheduler-dao/src/main/resources/org/apache/dolphinscheduler/dao/mapper/ProcessInstanceMapper.xml

```diff
@@ -37,6 +37,16 @@
         order by id asc
     </select>
 
+    <select id="queryTopNProcessInstance" resultType="org.apache.dolphinscheduler.dao.entity.ProcessInstance">
+        select *
+        from t_ds_process_instance
+        where state = 7
+        and start_time between
+        #{startTime} and #{endTime}
+        order by end_time-start_time desc
+        limit #{size}
+    </select>
```

Comment:
```suggestion
    <select id="queryTopNProcessInstance" resultType="org.apache.dolphinscheduler.dao.entity.ProcessInstance">
        select *
        from `t_ds_process_instance`
        where `state` = #{state}
        and `start_time` between
        #{startTime} and #{endTime}
        order by `end_time` - `start_time` desc
        limit #{size}
    </select>
```

Hi,

Good job, is it better to pass the `state` as a parameter to avoid the value change of state.
Thx a lot.

Reply:
I have 2 questions:
1.if we only want to query SUCCESS process instance,do we need a state parameter ?  If we need it,how do we put the parameter into the mapper function? 
2.why do we need  **`**   in this case since other cases don't use them?

Reply:
> I have 2 questions:
> 1.if we only want to query SUCCESS process instance,do we need a state parameter ? If we need it,how do we put the parameter into the mapper function?
> 2.why do we need **`** in this case since other cases don't use them?

Hi,
Thx a lot for your answer and your contribution.

For the first question, what I consider is to avoid changing the status code of `ExecutionStatus.SUCCESS`, for example, if you pass the code throw `ExecutionStatus.SUCCESS.getCode()`, it will avoid the error scene from changing the code of `ExecutionStatus.SUCCESS` from 7 to 6.
You can refer to [1] to look over the example of passing the parameter.

For the second question, using ``` is actually a good programming habit, which can avoid conflicts between SQL field parsing and keywords in SQL Engine.

[1] https://github.com/apache/incubator-dolphinscheduler/blob/dev/dolphinscheduler-dao/src/main/java/org/apache/dolphinscheduler/dao/mapper/ProcessInstanceMapMapper.java

Reply:
I still have a question about  the first previous question: since Java doesn't have default params,when I want to set SUCCESS as the default status,I might overload the method like this:
`    List<ProcessInstance> queryTopNProcessInstance(@Param("size") int size,
                                                   @Param("startTime") Date startTime,
                                                   @Param("endTime") Date endTime);`

 `   List<ProcessInstance> queryTopNProcessInstance(@Param("size") int size,
                                                   @Param("startTime") Date startTime,
                                                   @Param("endTime") Date endTime,
                                                   @Param("status")ExecutionStatus status);`
but since there is no method body here,how do I set SUCCESS as the default status?
I am not so familiar with Java,I hope you can give me some advice

Reply:
> I still have a question about the first previous question: since Java doesn't have default params,when I want to set SUCCESS as the default status,I might overload the method like this:
> ` List<ProcessInstance> queryTopNProcessInstance(@Param("size") int size, @Param("startTime") Date startTime, @Param("endTime") Date endTime);`
> 
> ` List<ProcessInstance> queryTopNProcessInstance(@Param("size") int size, @Param("startTime") Date startTime, @Param("endTime") Date endTime, @Param("status")ExecutionStatus status);`
> but since there is no method body here,how do I set SUCCESS as the default status?
> I am not so familiar with Java,I hope you can give me some advice

Hi,
Thx a lot for your answer and your contribution.

Can you tell me where you used `queryTopNProcessInstance` method except test part? This method was not used elsewhere according to your Pull Request.

According to your issue #3191 , if I understand it correctly, the purpose of this issue is to show the top-N tasks in the web interface. But at present, this PR does not seem to implement the issue completely. Are you still in the process of completing it?

----------------------
你可以告诉我在除了测试部分的其他哪些地方用到了 `queryTopNProcessInstance` 吗？根据你的 pr 来看，这个方法似乎没有被其他地方引用到。

根据你的 issue 来看，如果我理解正确的话，目的是想在web界面上展示运行时长topn的任务。但是目前这个 pr 似乎没有完全完成那个 issue，你是还在完成的过程中吗？

Reply:
I probably figured out how to make it: make status a param,and always call it with status SUCCESS(until I don't want to)